sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala

```diff
@@ -23,9 +23,54 @@ import org.apache.spark.sql.catalyst.analysis.TypeCheckResult.DataTypeMismatch
 import org.apache.spark.sql.catalyst.expressions.codegen._
 import org.apache.spark.sql.catalyst.expressions.codegen.Block._
 import org.apache.spark.sql.errors.QueryErrorsBase
-import org.apache.spark.sql.types.{AbstractDataType, DataType, StringType}
+import org.apache.spark.sql.types.{AbstractDataType, DataType, IntegerType, StringType}
 import org.apache.spark.unsafe.types.UTF8String
 
+/**
+ * The trait does Input Data Type validation .
+ */
+private[sql] trait Maskable extends QueryErrorsBase {
+
+  def validateInputDataTypes(
+      typeCheckResult: TypeCheckResult,
+      expressions: Seq[(Expression, String)],
+      additionalValidaton: () => Seq[Option[TypeCheckResult]] = () => Seq(None))
+      : TypeCheckResult = {
+
+    if (typeCheckResult.isSuccess) {
+      (expressions
+        .map { case (exp: Expression, message: String) =>
+          validateInputDataType(exp, message)
+        } ++ additionalValidaton()).flatten.headOption
+        .getOrElse(typeCheckResult)
+    } else {
+      typeCheckResult
+    }
+  }
+
+  private def validateInputDataType(exp: Expression, message: String): Option[TypeCheckResult] = {
+    if (!exp.foldable) {
+      Some(
+        DataTypeMismatch(
+          errorSubClass = "NON_FOLDABLE_INPUT",
+          messageParameters = Map(
+            "inputName" -> message,
+            "inputType" -> toSQLType(exp.dataType),
+            "inputExpr" -> toSQLExpr(exp))))
+    } else {
+      val replaceChar = exp.eval()
+      if (replaceChar != null && replaceChar.asInstanceOf[UTF8String].numChars != 1) {
```

Comment:
please use a pattern match against `exp.eval()` instead to (1) check that it is in fact an instance of `UTF8String` and (2) combine the new `if` to the previous `else`, de-denting the rest of the block?

Reply:
Done